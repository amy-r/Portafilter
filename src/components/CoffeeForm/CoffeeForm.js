import React, { Component } from 'react';
import { connect } from 'react-redux';
import { addCoffee } from '../../actions/index';
import { withRouter } from 'react-router';
import { addCoffeeData } from '../../Utilities/firebaseFunctions';
import { createEmail } from '../../Utilities/emailBody';
import '../Form/Form.css';
import PropTypes from 'prop-types';

export class CoffeeForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      name: '',
      email: this.props.email,
      overallScore: '',
      region: '',
      acidity: '',
      body: '',
      sweetness: '',
      tactile: '',
      overallImpression: '', 
      roaster: this.props.name,
      additionalComments: '',
      errorState: ''
    };
  }

  handleSubmit = async (event) => {
    try {
      event.preventDefault();
      const newCoffee = {...this.state};
      addCoffeeData(newCoffee);
      this.props.addCoffee(newCoffee);
      this.sendEmail(newCoffee);
      this.setState({
        name: '',
        email: '',
        overallScore: '',
        region: '',
        acidity: '',
        body: '',
        sweetness: '',
        tactile: '',
        overallImpression: '',
        roaster: '',
        additionalComments: ''
      });
      this.props.history.push(`/single-roaster/${this.props.name}`);
    } catch (error) {
      this.setState({
        errorState: 'Your email could not be sent at this time'
      });
    }
  }

  handleChange = (event) => {
    const {name, value} = event.target;
    this.setState({[name]:value});
  }

  sendEmail = async (form) => {
    try {
      const emailToSend = createEmail(form);
      await fetch('http://localhost:3001/send-form', {
        method: 'POST',
        headers: {
          'Accept': 'application/json',
          'Content-Type': 'application/json'
        },  
        body: JSON.stringify(emailToSend)
      });
    } catch (error) {
      this.setState({
        errorState: 'Your email could not be sent at this time'
      });
    }
  }

  render() {
    return (
      <div>
        <p> {this.state.errorState} </p>
        <form onSubmit= {this.handleSubmit} className='new-roaster'>
          <input type="text"
            className='full prefilled' 
            name="roaster" 
            value={this.props.name} 
            placeholder='Roaster Name'
            disabled
            required/>
          <input type="email"
            className='full prefilled' 
            name="email" 
            value={this.props.email}
            placeholder='Email' 
            disabled
            required/>
          <input type="text"
            className='full' 
            name="name" 
            value={this.state.name}
            placeholder='Blend/Bean Name' 
            onChange={this.handleChange}
            required/>
          <input type="text" 
            name="overallScore"
            className='full'  
            value={this.state.overallScore} 
            placeholder='Overall Score (1-6)'
            onChange={this.handleChange}
            required/>
          <input type="text" 
            className='full'
            name="region" 
            value={this.state.region} 
            placeholder='Region'
            onChange={this.handleChange}
            required/>
          <input type="text"
            className='full' 
            name="acidity" 
            value={this.state.acidity} 
            placeholder='Acidity (1-6)'
            onChange={this.handleChange}
            required/>
          <input type="text"
            className='full' 
            name="body" 
            value={this.state.body} 
            placeholder='Body (1-6)'
            onChange={this.handleChange}
            required/>
          <input type="text"
            className='full' 
            name="sweetness" 
            value={this.state.sweetness} 
            placeholder='Sweetness (1-6)'
            onChange={this.handleChange}
            required/>
          <input type="text"
            className='full' 
            name="tactile" 
            value={this.state.tactile} 
            placeholder='Tactile (1-6)'
            onChange={this.handleChange}
            required/>
          <input type="text"
            className='full' 
            name="overallImpression" 
            value={this.state.overallImpression} 
            placeholder='Overall Impression (Comments)'
            onChange={this.handleChange}
            required/>
          <input type="text" 
            className='full'
            name="additionalComments" 
            value={this.state.additionalComments} 
            placeholder='Additional Comments (Optional)'
            onChange={this.handleChange} />
          <input type="submit"className='submit'/>
        </form>
      </div>
    );
  }
}


export const mapDispatchToProps = (dispatch) => ({
  addCoffee: coffee => dispatch(addCoffee(coffee))
});

CoffeeForm.propTypes = {
  email: PropTypes.string,
  name: PropTypes.string,
  addCoffee: PropTypes.func,
  history: PropTypes.object
};

export default withRouter(connect(null, mapDispatchToProps)(CoffeeForm));
