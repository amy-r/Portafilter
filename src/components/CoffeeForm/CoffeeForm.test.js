import React from 'react';
import ReactDOM from 'react-dom';
import { CoffeeForm, mapDispatchToProps } from './CoffeeForm';
import { shallow } from 'enzyme';

describe('CoffeeForm', () => {
  it('should exist', () => {
    const wrapper = shallow(<CoffeeForm />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should start with a state of empty strings', () => {
    const expectedState = {
      name:'',
      overallScore: '',
      region: '',
      acidity: '',
      body: '',
      sweetness: '',
      tactile: '',
      overallImpression: '', 
      roaster: '',
      email: '',
      errorState:'',
      additionalComments: ''
    };

    const wrapper = shallow(<CoffeeForm />);

    expect(wrapper.state()).toEqual(expectedState);
  });

  it('should update state if the name field is typed into', () => {
    const expectedState = {
      name:'Jordan',
      overallScore: '',
      region: '',
      acidity: '',
      body: '',
      sweetness: '',
      tactile: '',
      overallImpression: '', 
      roaster: '',
      email: '',
      errorState: '',
      additionalComments: ''
    };
    const wrapper = shallow(<CoffeeForm />)
    const mockEvent = {
      target: {
        name: 'name',
        value: 'Jordan'
    }}

    wrapper.find('input').first().simulate('change', mockEvent);
    expect(wrapper.state()).toEqual(expectedState);
  })

  it('should update state of every input if they are modified', () => {
    const expectedState = {
      name:'1',
      overallScore: '1',
      region: '1',
      acidity: '1',
      body: '1',
      sweetness: '1',
      tactile: '1',
      email: '1',
      overallImpression: '1', 
      roaster: '1',
      errorState: '',
      additionalComments: '1'
    };

    const wrapper = shallow(<CoffeeForm />);

    const inputs = wrapper.find('input');
    const mockEventCreator = (inputName) => {
      return { 
        target: {
          name: inputName,
          value: '1'
        } 
      }
    };

    inputs.forEach( input => {
      input.simulate('change', mockEventCreator(input.prop('name')))
    });

    expect(wrapper.state()).toEqual(expectedState)
  });
})

describe('MDTP', () => {
  it('should call the dispatch function on MDTP', () => {
    const mockDispatch = jest.fn();
    const mapped = mapDispatchToProps(mockDispatch);
    mapped.addCoffee();
    expect(mockDispatch).toHaveBeenCalled;

  })
})