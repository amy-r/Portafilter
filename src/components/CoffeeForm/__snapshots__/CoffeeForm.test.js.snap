// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoffeeForm should exist 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CoffeeForm />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <input
          className="full"
          name="name"
          onChange={[Function]}
          placeholder="Name"
          type="text"
          value=""
/>,
        <input
          className="full"
          name="email"
          onChange={[Function]}
          placeholder="Email"
          type="email"
          value=""
/>,
        <input
          className="full"
          name="overallScore"
          onChange={[Function]}
          placeholder="Overall Score (1-6)"
          type="text"
          value=""
/>,
        <input
          className="full"
          name="region"
          onChange={[Function]}
          placeholder="Region"
          type="text"
          value=""
/>,
        <input
          className="full"
          name="acidity"
          onChange={[Function]}
          placeholder="Acidity (1-6)"
          type="text"
          value=""
/>,
        <input
          className="full"
          name="body"
          onChange={[Function]}
          placeholder="Body (1-6)"
          type="text"
          value=""
/>,
        <input
          className="full"
          name="sweetness"
          onChange={[Function]}
          placeholder="Sweetness (1-6)"
          type="text"
          value=""
/>,
        <input
          className="full"
          name="tactile"
          onChange={[Function]}
          placeholder="Tactile (1-6)"
          type="text"
          value=""
/>,
        <input
          className="full"
          name="overallImpression"
          onChange={[Function]}
          placeholder="Overall Impression (Comments)"
          type="text"
          value=""
/>,
        <input
          className="full"
          name="roaster"
          onChange={[Function]}
          placeholder="Roaster Name"
          type="text"
          value=""
/>,
        <input
          className="full"
          name="additionalComments"
          onChange={[Function]}
          placeholder="Additional Comments (Optional)"
          type="text"
          value=""
/>,
        <input
          className="submit"
          type="submit"
/>,
      ],
      "className": "new-roaster",
      "onSubmit": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "full",
          "name": "name",
          "onChange": [Function],
          "placeholder": "Name",
          "type": "text",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "full",
          "name": "email",
          "onChange": [Function],
          "placeholder": "Email",
          "type": "email",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "full",
          "name": "overallScore",
          "onChange": [Function],
          "placeholder": "Overall Score (1-6)",
          "type": "text",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "full",
          "name": "region",
          "onChange": [Function],
          "placeholder": "Region",
          "type": "text",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "full",
          "name": "acidity",
          "onChange": [Function],
          "placeholder": "Acidity (1-6)",
          "type": "text",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "full",
          "name": "body",
          "onChange": [Function],
          "placeholder": "Body (1-6)",
          "type": "text",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "full",
          "name": "sweetness",
          "onChange": [Function],
          "placeholder": "Sweetness (1-6)",
          "type": "text",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "full",
          "name": "tactile",
          "onChange": [Function],
          "placeholder": "Tactile (1-6)",
          "type": "text",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "full",
          "name": "overallImpression",
          "onChange": [Function],
          "placeholder": "Overall Impression (Comments)",
          "type": "text",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "full",
          "name": "roaster",
          "onChange": [Function],
          "placeholder": "Roaster Name",
          "type": "text",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "full",
          "name": "additionalComments",
          "onChange": [Function],
          "placeholder": "Additional Comments (Optional)",
          "type": "text",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "submit",
          "type": "submit",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
    ],
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <input
            className="full"
            name="name"
            onChange={[Function]}
            placeholder="Name"
            type="text"
            value=""
/>,
          <input
            className="full"
            name="email"
            onChange={[Function]}
            placeholder="Email"
            type="email"
            value=""
/>,
          <input
            className="full"
            name="overallScore"
            onChange={[Function]}
            placeholder="Overall Score (1-6)"
            type="text"
            value=""
/>,
          <input
            className="full"
            name="region"
            onChange={[Function]}
            placeholder="Region"
            type="text"
            value=""
/>,
          <input
            className="full"
            name="acidity"
            onChange={[Function]}
            placeholder="Acidity (1-6)"
            type="text"
            value=""
/>,
          <input
            className="full"
            name="body"
            onChange={[Function]}
            placeholder="Body (1-6)"
            type="text"
            value=""
/>,
          <input
            className="full"
            name="sweetness"
            onChange={[Function]}
            placeholder="Sweetness (1-6)"
            type="text"
            value=""
/>,
          <input
            className="full"
            name="tactile"
            onChange={[Function]}
            placeholder="Tactile (1-6)"
            type="text"
            value=""
/>,
          <input
            className="full"
            name="overallImpression"
            onChange={[Function]}
            placeholder="Overall Impression (Comments)"
            type="text"
            value=""
/>,
          <input
            className="full"
            name="roaster"
            onChange={[Function]}
            placeholder="Roaster Name"
            type="text"
            value=""
/>,
          <input
            className="full"
            name="additionalComments"
            onChange={[Function]}
            placeholder="Additional Comments (Optional)"
            type="text"
            value=""
/>,
          <input
            className="submit"
            type="submit"
/>,
        ],
        "className": "new-roaster",
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "full",
            "name": "name",
            "onChange": [Function],
            "placeholder": "Name",
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "full",
            "name": "email",
            "onChange": [Function],
            "placeholder": "Email",
            "type": "email",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "full",
            "name": "overallScore",
            "onChange": [Function],
            "placeholder": "Overall Score (1-6)",
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "full",
            "name": "region",
            "onChange": [Function],
            "placeholder": "Region",
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "full",
            "name": "acidity",
            "onChange": [Function],
            "placeholder": "Acidity (1-6)",
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "full",
            "name": "body",
            "onChange": [Function],
            "placeholder": "Body (1-6)",
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "full",
            "name": "sweetness",
            "onChange": [Function],
            "placeholder": "Sweetness (1-6)",
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "full",
            "name": "tactile",
            "onChange": [Function],
            "placeholder": "Tactile (1-6)",
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "full",
            "name": "overallImpression",
            "onChange": [Function],
            "placeholder": "Overall Impression (Comments)",
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "full",
            "name": "roaster",
            "onChange": [Function],
            "placeholder": "Roaster Name",
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "full",
            "name": "additionalComments",
            "onChange": [Function],
            "placeholder": "Additional Comments (Optional)",
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "submit",
            "type": "submit",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
      ],
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
